components/layout_2020/flow/float.rs

```diff
@@ -2,29 +2,484 @@
+        });
+        bands = bands.insert(FloatBand {
+            top: Length::new(f32::INFINITY),
+            left: None,
+            right: None,
+        });
+        FloatContext {
+            bands,
+            inline_size,
+            ceiling: Length::zero(),
+        }
+    }
+
+    /// Returns the current ceiling value. No floats may be placed (logically) above this line.
+    pub fn ceiling(&self) -> Length {
+        self.ceiling
+    }
+
+    /// (Logically) lowers the ceiling to at least `new_ceiling` units.
+    ///
+    /// If the ceiling is already logically lower (i.e. larger) than this, does nothing.
+    pub fn lower_ceiling(&mut self, new_ceiling: Length) {
+        self.ceiling = self.ceiling.max(new_ceiling);
+    }
+
+    /// Returns the highest block position that is both logically below the current ceiling and
+    /// clear of floats on the given side or sides.
+    pub fn clearance(&self, side: ClearSide) -> Length {
+        let mut band = self.bands.find(self.ceiling).unwrap();
+        while !band.is_clear(side) {
+            let next_band = self.bands.find_next(band.top).unwrap();
+            if next_band.top.px().is_infinite() {
+                break;
+            }
+            band = next_band;
+        }
+        band.top.max(self.ceiling)
+    }
+
+    /// Places a new float and adds it to the list. Returns the start corner of its margin box.
+    pub fn add_float(&mut self, new_float: FloatInfo) -> Vec2<Length> {
+        // Find the first band this float fits in.
+        let mut first_band = self.bands.find(self.ceiling).unwrap();
+        while !first_band.float_fits(&new_float, self.inline_size) {
+            let next_band = self.bands.find_next(first_band.top).unwrap();
+            if next_band.top.px().is_infinite() {
+                break;
+            }
+            first_band = next_band;
+        }
+
+        // The float fits perfectly here. Place it.
+        let (new_float_origin, new_float_extent);
+        match new_float.side {
+            FloatSide::Left => {
+                new_float_origin = Vec2 {
+                    inline: first_band.left.unwrap_or(Length::zero()),
+                    block: first_band.top.max(self.ceiling),
+                };
+                new_float_extent = new_float_origin.inline + new_float.size.inline;
+            },
+            FloatSide::Right => {
+                new_float_origin = Vec2 {
+                    inline: first_band.right.unwrap_or(self.inline_size) - new_float.size.inline,
+                    block: first_band.top.max(self.ceiling),
+                };
+                new_float_extent = new_float_origin.inline;
+            },
+        };
+        let new_float_rect = Rect {
+            start_corner: new_float_origin,
+            size: new_float.size,
+        };
+
+        // Split the first band if necessary.
+        first_band.top = new_float_rect.start_corner.block;
+        self.bands = self.bands.insert(first_band);
```

Comment:
Does insert() actually split the band? It seems to me that it just changes it

Reply:
Insert and split are the same operation, because bands only store their top value. e.g. if our bands have top values of 0px, 1px, 5px before, and we're inserting a float at y = 3px, then it becomes 0px, 1px, 3px, 5px after, effectively splitting the 1px band.